### ISSUE TITLE
Add an analyzer to prevent the explicit definition of the `EqualityContract` property, which establishes value-based equality in Records

### ISSUE THREAD

**Issue by egvijayanand** [dotnet/roslyn-analyzers#7586]
### Describe the problem you are trying to solve

Even though a reference type, Records exhibit Value-based equality.

For that, it generates the getter only `EqualityContract` property, which returns its actual type info (immutable).

However, this does not prevent it from being manually defined. It may be an intentional consideration.

To prevent unintentional usage, it would be advisable to implement an analyzer that alerts the user to its implications.

Run the below code for an example. The program is doing what is instructed to perform. Ideally, this should return `False` as types are incompatible.

I managed to get it working by returning the matching type info, but this can return any type info (say, `typeof(decimal`)) as it is not linked to the containing type. Hence, it is advisable to add this analyzer as soon as possible.

```cs
var origin = new Point2D(0, 0);
var initial = new Point3D(0, 0, 0);
Console.WriteLine(origin == initial);

record Point2D(int X, int Y)
{
    protected virtual Type EqualityContract => typeof(Point2D);
}

record Point3D(int X, int Y, int Z) : Point2D(X, Y)
{
    protected override Type EqualityContract => typeof(Point2D);
}
```

### Describe suggestions on how to achieve the rule

One possible way is when this `EqualityContract` is manually defined, validate whether the type returned is the same as the containing type.

### Additional context

Refer to the above code snippet.
